src/main/java/io/vertx/core/file/impl/FileResolver.java

```diff
@@ -375,11 +379,16 @@ private ClassLoader getClassLoader() {
     return cl;
   }
 
-  private synchronized void setupCacheDir(String id) {
-    String cacheDirName = fileCacheDir + "/file-cache-" + id;
-    cacheDir = new File(cacheDirName);
-    if (!cacheDir.mkdirs()) {
-      throw new IllegalStateException("Failed to create cache dir: " + cacheDirName);
+  private synchronized void setupCacheDir() {
+    FileAttribute<Set<PosixFilePermission>> ownerPermissions =
+        PosixFilePermissions.asFileAttribute(PosixFilePermissions.fromString("rwx------"));
+
+    try {
+      Path cacheDirBase = new File(cacheDirBaseName).toPath();
+      Files.createDirectories(cacheDirBase, ownerPermissions);
+      cacheDir = Files.createTempDirectory(cacheDirBase, "vertx-cache-", ownerPermissions).toFile();
+    } catch (IOException e) {
+      throw new UncheckedIOException(e.getMessage() + ": " + cacheDirBaseName, e);
```

Comment:
With this implementation it will no longer possible to run Vert.x on Windows if CP resolving is enabled.

Let's not deal with permissions here. If users need that level of confidentiality, there are a couple of options:

- use a dedicated `java.io.tmpdir` or `cacheDirBaseName`
- change the file creation mask before starting the JVM

Reply:
Thanks for catching this. I've added a check to only use PosixFilePermissions if the underlying file system supports posix attributes so that this also works on Windows.
I think that the file creation mask should not be used for this. The main goal of some vertx application might be to do some file tasks that can be configured by the file creation mask.
The vertx-cache has a completely different purpose, many sysops don't even know that it exists. Therefore the vertx-cache should have security by default by restricting to owner permissions. This is best practice, for example [mktemp](https://www.gnu.org/software/coreutils/manual/html_node/mktemp-invocation.html): 

> When creating a file, the resulting file has read and write permissions for the current user, but no permissions for the group or others